### PR TITLE
Fix ChannelCreator resolution broken #759

### LINKED ISSUE #759: ChannelCreator does not work in parent class
In #758, the ChannelCreator did not work, unless used in the last class defined, because the class dict contains only the items defined in that class, not inherited.

Does "inspect" help here?

### PR DESCRIPTION
**Pull request by bmoneke**
Fixes #759 

- adds a test to ensure that inheritance (and overriding as expected) works
- outsources the channel creation to its own method
- fixes the problem

Fix description:

1. Look through the MRO (in reverse order) and collect all `ChannelCreator` objects in a dictionary, the key is the variable name of the object. The reverse order ensures, that the `ChannelCreator` assigned to `channels` in a subclass replaces that of a superclass assigned also to `channels`.
2. Create the children according to that collected dictionary.

Not working possibilites:

- `inspect.getmembers` fails, because it tries to read all the attributes, even `Instrument.setting`, which is not possible.

### CODE REVIEW

File: pymeasure/instruments/common_base.py [hunk 1]
```diff
@@ -193,6 +188,24 @@ class attributes that are a DynamicProperty. Check also for class variables
...
 
+    def _create_channels(self):
+        """Create channels according to the ChannelCreator objects.
+
+        If a subclass overrides a ChannelCreator object with another one, the
+        one of the subclass is used, not that of the superclass.
+        """
+        channel_creators = {}
+        # Collect all ChannelCreators, override in order of subclasses
+        for obj in reversed((self,) + self.__class__.__mro__):
```
Comment (bilderbuchi) [inline pymeasure/instruments/common_base.py:199]: I'm nervous that we need to use `__mro__`, to be honest :sweat_smile:
The override should automatically be solved by a lookup -- since the MRO starts with the bottom-most class, if it finds a wanted attribute, it just stops. Why do we need to emulate this, here?

Second question, why are you prepending `self` to the mro list. On the one had, this seems to mix an instantiated object with the `<class>` objects that should be in the `__mro__` list. On the other hand, the class itself should already be the first entry in the MRO??

File: tests/instruments/test_common_base.py [hunk 2]
```diff
@@ -255,7 +255,26 @@ def test_remove_child_clears_attributes(self, parent_without_children):
...
+        function = CommonBase.ChannelCreator(GenericBase, "overridden", prefix=None)
+
+    @pytest.fixture()
+    def parent(self):
+        return self.InstrumentSubclass(ProtocolAdapter())
+
+    def test_inherited_children_are_present(self, parent):
+        assert isinstance(parent.ch_A, GenericBase)
+
+    def test_ChannelCreator_is_replaced(self, parent):
```
Comment (bilderbuchi) [inline tests/instruments/test_common_base.py:270]:
```suggestion
    def test_ChannelCreator_is_replaced_by_channel_collection(self, parent):
```
Maybe? Or add a comment that this is the usual mechanism -- I was initially confused why the changes from `InstrumentSubclass` would _replace_ anything in `.channels`.

File: pymeasure/instruments/common_base.py (hunk 1 above)
Comment (bilderbuchi) [reply pymeasure/instruments/common_base.py:199]: Blind code sketch what I mean:
```
            for item, value in self.__dict__.items():
                if isinstance(value, self.ChannelCreator):
                    channel_creators[item] = self.gettattr(value)  # let the MRO find the correct attribute (most derived class first)
```

Comment (bmoneke) [reply pymeasure/instruments/common_base.py:199]: If you give me a possibility to use automatic lookup, I'll use it.
The instance (`self`) dir does not contain them, and the class dirs do not inherit either.
Therefore, you have to look in the instance `self` (in case someone added it to the instance) , the class and every superclass (answering question 2).

I could use `dir()` and `getattr`, which would raise errors with `setting `.

File: tests/instruments/test_common_base.py (hunk 2 above)
Comment (bmoneke) [reply tests/instruments/test_common_base.py:270]: Good point.

File: pymeasure/instruments/common_base.py (hunk 1 above)
Comment (bmoneke) [reply pymeasure/instruments/common_base.py:199]: Neither `Instrument.setting` nor the special values of dynamic properties like to be read, but we could catch these errors with try clauses.

Comment (bmoneke) [reply pymeasure/instruments/common_base.py:199]: If you find such a possibility, we Don't need the dictionary anymore, but you can directly create the channels. I use the dictionary to search for the most recent definition before I add them to the instrument.

Comment (msmttchr) [reply pymeasure/instruments/common_base.py:199]: What about using inspect? It looks cleaner to me.
I think we have 3 places in the code base where such lookup is performed:
- parameters (using inspect)
- dynamic properties (using MRO)
- channel creator

I think once the best approach is selected, we could make the approach uniform.

Comment (bmoneke) [reply pymeasure/instruments/common_base.py:199]: `inspect.getmembers` fails, if you either have a dynamic property or a setting property, because the first one creates protected names (which cannot be read) and the latter cannot return a value either.

I based the code on that of dynamic properties, but I reversed the resolution order in order, that the most recent definition remains (as expected).

We could do it in the right order (from subclass to superclass), if we used `setdefault`, but I deemed direct assign better.

Comment (msmttchr) [reply pymeasure/instruments/common_base.py:199]: I think, that `inspect` doesn't fails, because we are reading attribute at class level, something like:
`for attr, value in inspect.getmembers(self.__class__):`
In this case, python properties are not evaluated and value returned is a property object.

Comment (bmoneke) [reply pymeasure/instruments/common_base.py:199]: Ah, getting the members of the class is great, thanks. I'll try it at home.

Shall I change special names as well to use inspect?

Comment (bilderbuchi) [reply pymeasure/instruments/common_base.py:199]: Make sense to me.

Comment (msmttchr) [reply pymeasure/instruments/common_base.py:199]: > Shall I change special names as well to use inspect?

Yes, please. Also take into account that special names uses also `self` in the attribute lookup, which means that take into account also instance, but I believe this is a bug and so search in class only is also fixing this.

Comment (bmoneke) [reply pymeasure/instruments/common_base.py]: Change is implemented (together with additional tests, to verify, that dynamic properties continue to work as expected).

File: tests/instruments/test_common_base.py [hunk 3]
```diff
@@ -114,7 +114,14 @@ def fake():
...
     # Keep values unchanged, just derive another instrument, e.g. to add more properties
-    pass
+    fake_ctrl2 = CommonBase.control(
+        "", "%d", "docs",
+        validator=truncated_range,
+        values=(1, 10),
+        dynamic=True,
+    )
+
+    fake_ctrl2.values = (2, 20)
```
Comment (msmttchr) [inline tests/instruments/test_common_base.py:124]:
```suggestion
    fake_ctrl2_values = (2, 20)
```
Shouldn't be like this ?

**Review by msmttchr — CHANGES_REQUESTED** [review]
@bmoneke looks alright to me, except one comment.

File: tests/instruments/test_common_base.py (hunk 3 above)
Comment (bmoneke) [reply tests/instruments/test_common_base.py:124]: Yes...

I just verified, that all tests pass with this change.

**Review by msmttchr — APPROVED** [review]
It is OK for me. Thanks for your work (both quantity and quality)

File: tests/instruments/test_common_base.py (hunk 3 above)
Comment (bilderbuchi) [reply tests/instruments/test_common_base.py]: Funny that the tests passed before, too... is that something we should investigate?

File: tests/instruments/test_common_base.py [hunk 4]
```diff
@@ -711,3 +719,18 @@ def test_dynamic_property_values_update_in_one_instance_leaves_other_unchanged()
...
+    with pytest.raises(AttributeError):
+        inst.fake_ctrl_validator
+    # Test for new attribute
+    with pytest.raises(AttributeError):
+        inst.fake_ctrl2_validator
+
+
+def test_dynamic_property_values_defined_at_superclass_level():
+    inst = StrictExtendedBase()
+    inst.fake_ctrl2 = 17  # outside original values (1, 10)
```
Comment (bilderbuchi) [inline tests/instruments/test_common_base.py:736]: This test doesn't assert anything?

Comment (bmoneke) [reply tests/instruments/test_common_base.py:736]: The assignment would fail, if the values parameter would not have been changed from the original values (1, 10) to (2, 20).
Not failing is passing.
And I failed in making the comment understandable...

File: tests/instruments/test_common_base.py (hunk 3 above)
Comment (bmoneke) [reply tests/instruments/test_common_base.py]: I already thought about it. I used "truncated" instead of "strict".
Now I fixed it and expanded the test.

**Review by bilderbuchi — APPROVED** [review]
LGTM. :shipit:
